clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1521,9 +1522,13 @@ event_exit(void)
 #ifdef BUILD_PT_TRACER
     if (op_offline.get_value() && op_enable_kernel_tracing.get_value()) {
         drpttracer_exit();
-        /* Dump kcore and kallsyms to {kernel_pt_logsubdir}. */
+        /* Dump kcore to kimage and kimage.metadata, and store the two file to
+         * {kernel_pt_logsubdir}.
+         * The Kernel dump function is only invoked at processes exit on the Linux X86_64
+         * platform.
+         */
         if (!syscall_pt_trace_t::kernel_image_dump(kernel_pt_logsubdir)) {
-            NOTIFY(0, "WARNING: failed to run shellscript to dump kernel image\n");
+            NOTIFY(0, "WARNING: failed to dump kernel image\n");
```

Comment:
Just like for raw2trace, tracer can't directly write to files: we need indirection to support library-based filesystems.  This file dumping needs to go through the write_file indirection.